server.py

```diff
@@ -1,18 +1,145 @@
+    else:
+        return part_list[1]
+
+
+def response_ok(path, uri, file_type):
+    """Create customized HTTP response for route_uri()."""
+    return_path = os.path.join(path, uri)
+    if os.path.isfile(return_path) is False:
+        return four_oh_four()
+    else:
+        if "image" in file_type:
+            open_file = io.open(return_path, 'rb')
+        else:
+            open_file = io.open(return_path)
+        size = os.path.getsize(return_path)
+        read_file = open_file.read()
+        open_file.close()
+        html_response = "HTTP/1.1 200 OK\r\nContent-Type: {}\r\nContent-length: {}\r\n\r\n".format(file_type, size) + read_file
+        return html_response
+
+
+def four_oh_four():
+    """Send 404 not found message."""
+    open_file = io.open("/Users/admin-1/http_server/http-server/webroot/images/404.jpg" , "rb")
+    read_file = open_file.read()
+    open_file.close()
+    html_response = "HTTP/1.1 400 Not Found\r\n Content-Type: image/jpeg\r\n\r\n" + read_file
+    return html_response
+
+
+def generate_html(base_path, uri):
+    """Create html code to display files."""
+    if uri == "/" or uri == "":
+        join_path = base_path
+    else:
+        join_path = os.path.join(base_path, uri)
+    path = os.walk(join_path)
+    walk_path = next(path)
+    show_dir = walk_path[1]
+    show_files = walk_path[2]
+    inject_html_dir = ""
+    inject_html_file = ""
+    # Create html <li> elements of contents of folder
+    if len(show_dir) > 0:
+        for directory in show_dir:
+            inject_html_dir += "<li><a href={}>{}</a></li>".format(os.path.join(uri, directory), directory)
+    else:
+        inject_html_dir = "No Directories"
+    if len(show_files) > 0:
+        for file in show_files:
+            inject_html_file += "<li><a href={}>{}</a></li>".format(os.path.join(uri, file), file)
+    else:
+        inject_html_file = "No Files"
+    return_path = "Directories:<ul>{}</ul></br>Files:<ul>{}</ul>".format(inject_html_dir, inject_html_file)
+    html_response = "HTTP/1.1 200 OK\r\nContent-Type: text/html\r\n\r\n" + return_path
+    print(html_response)
+    return html_response
+
+
+def route_uri(base_path, uri):
+    """Check to see what file type GET is looking for."""
+    # We have to remove the FIRST slash in the requested uri or else it throws off os.path.join()
+    uri_no_slash = uri.replace("/", "", 1)
+    if "." not in uri:
+        return generate_html(base_path, uri_no_slash)
+    elif ".html" in uri:
+        return response_ok(base_path, uri_no_slash, "text/html")
+    elif ".py" in uri:
+        return response_ok(base_path, uri_no_slash, "text/py")
+    elif ".txt" in uri:
+        return response_ok(base_path, uri_no_slash, "text/plain")
+    # images
+    elif ".jpg" in uri:
+        return response_ok(base_path, uri_no_slash, "image/jpeg")
+    elif ".png" in uri:
+        return response_ok(base_path, uri_no_slash, "image/png")
+    else:
+        return four_oh_four()
```

Comment:
this is only good for the file types that are already there.  There is a way to figure out the same information in a way that is not so limited.

```python
import mimetypes
import os

extension = os.path.splitext(uri_no_slash)
mimetype = mimetypes.types_map(extension)
```
This way you don't need to special case all the different types, you can get all this method has plus more in three lines of code, and you only have to call response_ok once.

You can also look a bit deeper into the os.path module to find tools that will let you test to see if the path you have is a directory or a file.  There are a lot of tools there, and I expect you to look at the documentation to find them